src/core/ir/discriminator.rs

```diff
@@ -0,0 +1,1054 @@
+use std::collections::HashSet;
+use std::fmt::Write;
+use std::ops::{Deref, DerefMut};
+
+use anyhow::{bail, Result};
+use async_graphql::Value;
+use indenter::indented;
+use indexmap::IndexMap;
+
+use crate::core::config::Type;
+use crate::core::valid::{Cause, Valid, Validator};
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub enum TypeName {
```

Comment:
Can you add more documentation about what each type does and the overall algorithm.

Reply:
added more documentation